sql/auth/sql_user.cc

```diff
@@ -2864,6 +2886,146 @@ static bool check_orphaned_definers(THD *thd, List<LEX_USER> &list) {
+ warning.
+
+ 1. Source: ER_AUTHENTICATION_POLICY_MISMATCH warning here => TOI
+    Replica: No warnings, no errors
+    Inconsistency voting: not triggered
+ 2. Source: ER_AUTHENTICATION_POLICY_MISMATCH error here => no TOI, continue
+            flow, return with ER_AUTHENTICATION_POLICY_MISMATCH
+    Inconsistency voting: not triggered
+ 3. Source: ER_AUTHENTICATION_POLICY_MISMATCH warning here => do TOI then return
+            with ER_PLUGIN_IS_NOT_LOADED
+    Replica: error with ER_PLUGIN_IS_NOT_LOADED
+    Inconsistency voting: triggered for both sides with ER_PLUGIN_IS_NOT_LOADED
+ */
+class Silence_any_error final : public Internal_error_handler {
+ public:
+  bool handle_condition(THD *, uint, const char *,
+                        Sql_condition::enum_severity_level *,
+                        const char *) override {
+    return true;
+  }
+};
+static bool wsrep_check_for_auth_policy(THD *thd, List<LEX_USER> &list,
+                                        const char *cmd) {
+  if (WSREP(thd) && !thd->wsrep_applier && list.size() > 0) {
+    Table_ref tables[ACL_TABLES::LAST_ENTRY];
+    bool transactional_tables;
+
+    if (open_grant_tables(thd, tables, &transactional_tables, true)) {
+      return true;
+    }
+
+    raii::Sentry<> close_acl_tables_guard{
+        [thd]() -> void { commit_and_close_mysql_tables(thd); }};
+
+    // Get WRITE_MODE lock, because we will access ACL cache. It may happen
+    // that entry is not yet in cache and needs to be added.
+    Acl_cache_lock_guard acl_cache_lock(thd, Acl_cache_lock_mode::WRITE_MODE);
+
+    if (!acl_cache_lock.lock()) {
+      return true;
+    }
+
+    LEX_USER *user_from, *tmp_user_from;
+    List_iterator<LEX_USER> user_list(list);
+    Silence_any_error error_handler;
+    bool is_privileged_user = is_privileged_user_for_credential_change(thd);
+
+    thd->push_internal_handler(&error_handler);
+    raii::Sentry<> error_handler_pop_guard{
+        [thd]() -> void { thd->pop_internal_handler(); }};
+
+    while ((tmp_user_from = user_list++)) {
+      acl_table::Pod_user_what_to_update what_to_alter;
+      bool history_check_done = false;
+      I_multi_factor_auth *mfa = nullptr;
+      Userhostpassword_list generated_passwords;
+
+      if (!(user_from = get_current_user(thd, tmp_user_from))) {
+        // let the original flow do the job
+        return false;
+      }
+
+      /* set_and_validate_user_attributes() will hash provided password and
+       wipe-out its plaintext version in
+       user_from_tmp.first_factor_auth_info.auth.str.
+       LEX_CSTRING holds just a pointer, so we need to do a deep copy of
+       LEX_USER for original flow to work properly. */
+      LEX_USER *user_from_tmp = deep_copy(thd, user_from);
+
+      /* Skip password verification. It would append keys to writeset, but TOI
+       is not started yet, so it is not allowed. We don't need it now. Original
+       flow will handle it */
+      if (set_and_validate_user_attributes(
```

Comment:
Taking into account that we skip password verification here anyway, do we really need to pass Table_ref * which corresponds to really open password table here? Perhaps we can pass nullptr or pointer to dummy object instead?
That would make all the opening of privilege tables above unnecessary. Removing latter would make changes to open_grant_tables() unnecessary in its turn as far as I can see.

Reply:
I just copied the validation code from the original flow that follows. But you are correct. As we don't check password history, opening tables is not needed, so we can skip this step and pass nullptr as a history table.